packages/legend-query/src/components/QueryBuilderValueSpecificationEditor.tsx

```diff
@@ -534,75 +599,601 @@ export const LatestDatePrimitiveInstanceValueEditor: React.FC = () => (
+          ).map((t) => ({
+            value: t.toString(),
+            label: t.toString(),
+          }))}
+          onChange={(startDate: CustomDateInputOption): void => {
+            setStartDateValue(startDate.value);
+            updateValue(
+              durationRangeValue,
+              durationUnitValue,
+              adjustmentValue,
+              startDate.value,
+            );
+          }}
+          value={{ value: startDateValue, label: startDateValue }}
+          darkMode={true}
+        />
+      </div>
+    </div>
+  );
+};
+
+export const CustomStartDayOfDateValueSpecificationEditor: React.FC<{
+  graph: PureModel;
+  customDateAdjustOptionValue: CustomDateOption;
+  updateValueSpecification: (val: ValueSpecification) => void;
+  setDateOptionValue: Dispatch<SetStateAction<CustomDateOption>>;
+}> = (props) => {
+  const {
+    graph,
+    customDateAdjustOptionValue,
+    updateValueSpecification,
+    setDateOptionValue,
+  } = props;
+
+  const [durationUnitValue, setDurationUnitValue] = useState(
+    customDateAdjustOptionValue.unit &&
+      Object.values(QUERY_BUILDER_DATE_DURATION_UNIT).filter(
+        (c) => c.toString() === customDateAdjustOptionValue.unit,
+      ).length > 0
+      ? customDateAdjustOptionValue.unit
+      : '',
+  );
+  const updateValue = (latestDurationUnitValue: string): void => {
+    if (latestDurationUnitValue !== '') {
+      const startDayOfDateOption = generateDateOptionForStartDayOfDateOption(
+        latestDurationUnitValue,
+      );
+      updateValueSpecification(
+        generateDateFunctionExpression(graph, startDayOfDateOption),
+      );
+      setDateOptionValue({
+        label: startDayOfDateOption.label,
+        value: QUERY_BUILDER_DATE_OPTION.START_DAY_OF_DATE,
+        unit: latestDurationUnitValue,
+      });
+    }
+  };
+
+  return (
+    <div className="query-builder-value-spec-editor__date-option-dropdown__custom-date">
+      <div className="query-builder-value-spec-editor__date-option-dropdown__custom-date__input">
+        <div className="query-builder-value-spec-editor__date-option-dropdown__custom-date__input-label">
+          When
+        </div>
+        <CustomSelectorInput
+          className="query-builder-value-spec-editor__date-option-dropdown__custom-date__input-dropdown"
+          options={Object.values(QUERY_BUILDER_DATE_DURATION_UNIT).map((t) => ({
+            value: t.toString(),
+            label: t.toString(),
+          }))}
+          onChange={(durationUnit: CustomDateInputOption): void => {
+            setDurationUnitValue(durationUnit.value);
+            updateValue(durationUnit.value);
+          }}
+          value={{ value: durationUnitValue, label: durationUnitValue }}
+          darkMode={true}
+        />
+      </div>
+    </div>
+  );
+};
+
+export const CustomPreviousDayOfWeekValueSpecificationEditor: React.FC<{
+  graph: PureModel;
+  customDateAdjustOptionValue: CustomDateOption;
```

Comment:
let's reorder `customDateAdjustOptionValue` to be the first param. Sorry for this nitpickiness :pray: